crates/router/src/core/payment_methods/network_tokenization.rs

```diff
@@ -0,0 +1,700 @@
+    state: &routes::SessionState,
+    payload_bytes: &[u8],
+    amount: String,
+    currency: String,
+    customer_id: id_type::CustomerId,
+) -> CustomResult<(CardNetworkTokenResponsePayload, Option<String>), errors::NetworkTokenizationError>
+{
+    let tokenization_service = &state.conf.network_tokenization_service.get_inner();
+    let enc_key = tokenization_service.public_key.peek().clone();
+
+    let key_id = tokenization_service.key_id.clone();
+
+    let jwt = encryption::encrypt_jwe(
+        payload_bytes,
+        enc_key,
+        EncryptionAlgorithm::A128GCM,
+        Some(key_id.as_str()),
+    )
+    .await
+    .change_context(errors::NetworkTokenizationError::SaveNetworkTokenFailed)
+    .attach_printable("Error on jwe encrypt")?;
+
+    let order_data = OrderData {
+        consent_id: uuid::Uuid::new_v4().to_string(),
+        customer_id,
+        amount,
+        currency,
+    };
+
+    let api_payload = ApiPayload {
+        service: "NETWORK_TOKEN".to_string(),
+        card_data: jwt,
+        order_data,
+        key_id,
+        should_send_token: true,
+        sub_merchant_id: "visa_sbx_working".to_string(),
+    };
+
+    let mut request = services::Request::new(
+        services::Method::Post,
+        tokenization_service.generate_token_url.as_str(),
+    );
+    request.add_header(headers::CONTENT_TYPE, "application/json".into());
+    request.add_header(
+        headers::AUTHORIZATION,
+        tokenization_service
+            .token_service_api_key
+            .peek()
+            .clone()
+            .into_masked(),
+    );
+    request.set_body(RequestContent::Json(Box::new(api_payload)));
+
+    let response = services::call_connector_api(state, request, "generate_token")
+        .await
+        .change_context(errors::NetworkTokenizationError::ResponseDeserializationFailed);
+
+    let res = response
+        .map_err(|error| {
+            error.change_context(errors::NetworkTokenizationError::ResponseDeserializationFailed)
+        })
+        .attach_printable("Error while receiving response")
+        .and_then(|inner| match inner {
+            Err(err_res) => {
+                let parsed_error: NetworkTokenErrorResponse = err_res
+                    .response
+                    .parse_struct("Card Network Tokenization Response")
+                    .change_context(
+                        errors::NetworkTokenizationError::ResponseDeserializationFailed,
+                    )?;
+                logger::error!(
+                    error_code = %parsed_error.error_info.code,
+                    developer_message = %parsed_error.error_info.developer_message,
+                    "Network tokenization error: {}",
+                    parsed_error.error_message
+                );
+                Err(errors::NetworkTokenizationError::ResponseDeserializationFailed)
+                    .attach_printable(format!("Response Deserialization Failed: {err_res:?}"))
```

Comment:
You can use `change_context` instead. Optional change, by the way.

```suggestion
                Err(err_res)
                    .change_context(errors::NetworkTokenizationError::ResponseDeserializationFailed)
```

Reply:
I can't create a Report of `err_res` since it doesn't satisfy the required traits